scheduler/src/cook/mesos/compute_cluster.clj

```diff
@@ -0,0 +1,105 @@
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.mesos.compute-cluster
+  (:require         [clojure.tools.logging :as log]
+                    [datomic.api :as d :refer (q)]))
+
+(defn- create-missing-compute-cluster
+  "Create a missing compute-cluster for one thats not yet in the database."
+  [conn compute-cluster]
+  (log/info "Installing a new compute cluster in datomic for " compute-cluster)
+  (d/transact
+    conn
+    [(assoc compute-cluster :db/id (d/tempid :db.part/user))]))
+
+(defn- mesos-cluster->compute-cluster-map-for-datomic
+  "Given a mesos cluster dictionary, determine the datomic entity it should correspond to."
+  [{:keys [compute-cluster-name framework-id]}]
+  {:compute-cluster/type :compute-cluster.type/mesos
+   :compute-cluster/cluster-name compute-cluster-name
+   :compute-cluster/mesos-framework-id framework-id})
+
+(defn get-mesos-cluster-entity
+  "Given a configuration map for a mesos cluster, return the datomic entity corresponding to the cluster, if it exists. Internal helper function."
+  [unfiltered-db {:keys [compute-cluster-name framework-id]}]
+  {:pre [compute-cluster-name
+         framework-id]}
+  (let [query-result
+        (q '[:find [?c]
+             :in $ ?cluster-name? ?mesos-id?
+             :where
+             [?c :compute-cluster/type :compute-cluster.type/mesos]
+             [?c :compute-cluster/cluster-name ?cluster-name?]
+             [?c :compute-cluster/mesos-framework-id ?framework-id?]]
+           unfiltered-db compute-cluster-name framework-id)
+        materialized-query-result (into [] query-result)]
+    materialized-query-result))
+
+(defn get-mesos-cluster-map
+  "Process one mesos cluster specification, returning the entity id of the corresponding compute-cluster, creating the cluster if it does not exist."
+  [conn {:keys [compute-cluster-name framework-id] :as mesos-cluster}]
+  {:pre [compute-cluster-name
+         framework-id]}
+  (let [cluster-entity (get-mesos-cluster-entity (d/db conn) mesos-cluster)]
+    (when (zero? (count cluster-entity))
+      (create-missing-compute-cluster conn (mesos-cluster->compute-cluster-map-for-datomic mesos-cluster)))
+    {:cluster-type :mesos-cluster
+     :compute-cluster-name compute-cluster-name
+     :mesos-framework-id framework-id
+     :db-id (->> (get-mesos-cluster-entity (d/db conn) mesos-cluster)
+                 first)}))
+
+(defn get-mesos-clusters-from-config
+  "Get all of the mesos clusters defined in the configuration.
+  This is a wart that undo'ed the shattering of the :mesos {...} into separate keys that
+  occurs in config.edn, and consolidates it into one dictionary for downstream consumption."
+  [{:keys [mesos-compute-cluster-name mesos-framework-id]}]
+  [{:compute-cluster-name mesos-compute-cluster-name :framework-id mesos-framework-id}])
+
+(def cluster-name->cluster-dict-atom (atom nil))
+
+(defn cluster-name->db-id
```

Comment:
The fact that we're using db ids to refer to these is kind of smelly to me. I get that we don't have a unique key for each cluster, but what about using some kind of composite key? Maybe something like `mesos:/dft/Cook_dft` or something like that. Don't expose it, just purely as an implementation detail. But some way to go from name/framework-id -> cluster.